### DISCUSSION TITLE
Extracting Deviance Explained from tinyVAST

### DISCUSSION THREAD

**Discussion by PeriGerson** [vast-lib/tinyVAST#54] [Q&A]
Hi Jim,

I am working on a multivariate model in tinyVAST and would like to compare it to stacked univariate models. We discussed using deviance explained as a comparison point. Would you mind explaining how to extract the deviance explained from tinyVAST for this kind of comparison? I have included example code for the multivariate and univariate models below. Thank you!

\# multivariate model
 
\# define covariance/variance structure
sem = "
walleyepollock <-> walleyepollock, pollock_sd
capelin <-> capelin, capelin_sd
sandlance <-> sandlance, sandlance_sd
pacific_herring <-> pacific_herring, herring_sd
walleyepollock <-> capelin, pollock_capelin
walleyepollock <-> sandlance, pollock_sandlance
walleyepollock <-> pacific_herring, pollock_herring
capelin <-> sandlance, capelin_sandlance
capelin <-> pacific_herring, capelin_herring
sandlance <-> pacific_herring, sandlance_herring
"
 
 
\# multivariate tinyVAST model
 
multi = tinyVAST( data = foragefish_data,
                    formula = prey_wt ~ s(depth) +
                      s(pred_length) +
                      s(bottom_temp, by = as.factor(prey_taxa)) +
                      factor(year) +
                      factor(predator) +
                      factor(prey_taxa),
                    family = tweedie(),
                    spatial_graph = mesh,
                    variable_column = "prey_taxa",
                    sem = sem)
 
 
\#univariate model example
 
\# walleye pollock weight as response
\# define variance structure
 
sem = "
walleyepollock <-> walleyepollock, pollock_sd
"
 
\# pollock univariate model
 
pollock = tinyVAST( data = pollock_data,
                    formula = prey_wt ~ s(depth) +
                      s(pred_length) +
                      s(bottom_temp) +
                      factor(year) +
                      factor(predator),
                    family = tweedie(),
                    spatial_graph = mesh,
                    variable_column = "prey_taxa",
                    sem = sem)

**Comment by James-Thorson-NOAA** [comment]
and can you confirm that the univariate model has a slot `pollock@deviance_explained` and the multivariate one has slot `multi@deviance_explained`?  

I'm assuming it does and you're asking because of the note in `?deviance_explained` about multivariate models.  In that case, I recommend running 
```R
R2 = deviance_explained( multi, null_formula = prey_wt ~ [something factor])
```

However, I'm guessing you want an intercept for every prey:predator, so editing the JSDM to:

```R
multi = tinyVAST( data = foragefish_data,
formula = prey_wt ~ s(depth) +
s(pred_length) +
s(bottom_temp, by = as.factor(prey_taxa)) +
factor(year) +
interaction(predator,prey_taxa) +
family = tweedie(),
spatial_graph = mesh,
variable_column = "prey_taxa",
sem = sem)
```

In which case the `null_formula = prey_wt ~ interaction(predator,prey_taxa)` makes sense to me.  The point is that you want the null formula to match whatever null assumption you'd make.  In this case, I'm assuming you'd compare it with a separate SDM for each combination of predator and prey.

I'm happy to keep discussing, thanks for bring it up.  Closing the tab for now.

**Reply by PeriGerson** [reply]
Thanks Jim!

Yes, the univariate and multivariate models have slots for the deviance_explained and your assumption was correct.

Would I then compare the deviance explained using the null assumption of the predator-prey interaction term to an aggregate of the deviance explained of the SDMs for each predator and prey combo?

**Comment by James-Thorson-NOAA** [comment]
I'm not really following the wording, but you could:

1. run `deviance_explained` manually while specifying that alternative `null_formula` to get the R2 for the JSDM with a single function call, OR
2. run a separate null (intercept-only) SDM for each species-prey combination, compute the sum of relative deviance across those, and recompute the R2 yourself

I think #1 sounds easier.

**Reply by PeriGerson** [reply]
Ahh okay, I was confused about the interpretation of the R2 from the deviance_explained function. Thanks!
